From e7c6365bdff2bb75bc6265cdfaec0a20ecebaf2a Mon Sep 17 00:00:00 2001
From: Fei Wang <fei.w.wang@intel.com>
Date: Tue, 14 Jul 2020 10:21:55 +0800
Subject: [PATCH 06/51] lavc/vaapi_hevc: support 420/422/444 12bit enc/dec

---
 libavcodec/vaapi_decode.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/libavcodec/vaapi_decode.c b/libavcodec/vaapi_decode.c
index 134f10eca5..7f68e839d0 100644
--- a/libavcodec/vaapi_decode.c
+++ b/libavcodec/vaapi_decode.c
@@ -301,6 +301,7 @@ static int vaapi_decode_find_best_format(AVCodecContext *avctx,
     VAStatus vas;
     VASurfaceAttrib *attr;
     enum AVPixelFormat source_format, best_format, format;
+    const AVPixFmtDescriptor *desc, *desc_s;
     uint32_t best_fourcc, fourcc;
     int i, j, nb_attr;
 
@@ -348,6 +349,13 @@ static int vaapi_decode_find_best_format(AVCodecContext *avctx,
         av_log(avctx, AV_LOG_DEBUG, "Considering format %#x -> %s.\n",
                fourcc, av_get_pix_fmt_name(format));
 
+        if (best_format != AV_PIX_FMT_NONE) {
+            desc = av_pix_fmt_desc_get(format);
+            desc_s = av_pix_fmt_desc_get(source_format);
+            if (desc->comp[0].depth != desc_s->comp[0].depth)
+                continue;
+        }
+
         best_format = av_find_best_pix_fmt_of_2(format, best_format,
                                                 source_format, 0, NULL);
         if (format == best_format)
-- 
2.25.1

