From 874cf3ce915a7abec762d584364b0bd690fc6645 Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Tue, 1 Dec 2020 10:32:58 +0800
Subject: [PATCH 52/81] avframe: zero the frame, including padding data

---
 libavutil/frame.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavutil/frame.c b/libavutil/frame.c
index fefb2b69c0..21118c782d 100644
--- a/libavutil/frame.c
+++ b/libavutil/frame.c
@@ -256,7 +256,7 @@ static int get_video_buffer(AVFrame *frame, int align)
         total_size += sizes[i];
     }
 
-    frame->buf[0] = av_buffer_alloc(total_size);
+    frame->buf[0] = av_buffer_allocz(total_size);
     if (!frame->buf[0]) {
         ret = AVERROR(ENOMEM);
         goto fail;
-- 
2.25.1

