From cb3c29ca86543ed34b53dff3fa8173d65a4d9d2f Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Tue, 1 Dec 2020 10:32:58 +0800
Subject: [PATCH 52/82] avframe: zero the frame, including padding data

---
 libavutil/frame.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavutil/frame.c b/libavutil/frame.c
index 2ec59b44b175..e767472c45a4 100644
--- a/libavutil/frame.c
+++ b/libavutil/frame.c
@@ -163,7 +163,7 @@ static int get_video_buffer(AVFrame *frame, int align)
         total_size += sizes[i];
     }
 
-    frame->buf[0] = av_buffer_alloc(total_size);
+    frame->buf[0] = av_buffer_allocz(total_size);
     if (!frame->buf[0]) {
         ret = AVERROR(ENOMEM);
         goto fail;
-- 
2.25.4

