From b9c0f03b7203323f768ddf5663ddd1c8e7dcdc0c Mon Sep 17 00:00:00 2001
From: Kizna1ver <jianyuliucsu@gmail.com>
Date: Wed, 14 Sep 2022 17:22:27 +0800
Subject: [PATCH 3/3] dnn/vf_dnn_classify.c: add AV_PIX_FMT_VAAPI format
 support.

---
 libavfilter/vf_dnn_classify.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavfilter/vf_dnn_classify.c b/libavfilter/vf_dnn_classify.c
index d242aebcfb..887d1f25ea 100644
--- a/libavfilter/vf_dnn_classify.c
+++ b/libavfilter/vf_dnn_classify.c
@@ -198,7 +198,7 @@ static const enum AVPixelFormat pix_fmts[] = {
     AV_PIX_FMT_GRAY8, AV_PIX_FMT_GRAYF32,
     AV_PIX_FMT_YUV420P, AV_PIX_FMT_YUV422P,
     AV_PIX_FMT_YUV444P, AV_PIX_FMT_YUV410P, AV_PIX_FMT_YUV411P,
-    AV_PIX_FMT_NV12,
+    AV_PIX_FMT_NV12, AV_PIX_FMT_VAAPI,
     AV_PIX_FMT_NONE
 };
 
-- 
2.25.1

