From 708fcc2e789cd687982876fc286ade58377c9d27 Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Wed, 8 Dec 2021 12:35:16 +0800
Subject: [PATCH 93/95] lavc/qsvenc: add VP9 profiles

---
 libavcodec/qsvenc.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/libavcodec/qsvenc.c b/libavcodec/qsvenc.c
index cf870e0d9f..1762be2822 100644
--- a/libavcodec/qsvenc.c
+++ b/libavcodec/qsvenc.c
@@ -79,6 +79,15 @@ static const struct profile_names hevc_profiles[] = {
 #endif
 };
 
+static const struct profile_names vp9_profiles[] = {
+#if QSV_VERSION_ATLEAST(1, 19)
+    { MFX_PROFILE_VP9_0,                        "0"                     },
+    { MFX_PROFILE_VP9_1,                        "1"                     },
+    { MFX_PROFILE_VP9_2,                        "2"                     },
+    { MFX_PROFILE_VP9_3,                        "3"                     },
+#endif
+};
+
 static const char *print_profile(enum AVCodecID codec_id, mfxU16 profile)
 {
     const struct profile_names *profiles;
@@ -100,6 +109,11 @@ static const char *print_profile(enum AVCodecID codec_id, mfxU16 profile)
         num_profiles = FF_ARRAY_ELEMS(hevc_profiles);
         break;
 
+    case AV_CODEC_ID_VP9:
+        profiles = vp9_profiles;
+        num_profiles = FF_ARRAY_ELEMS(vp9_profiles);
+        break;
+
     default:
         return "unknown";
     }
-- 
2.17.1

