From 83eac25c838246e14a25f6727c309169f7cde1f6 Mon Sep 17 00:00:00 2001
From: Wenbin Chen <wenbin.chen@intel.com>
Date: Tue, 5 Mar 2024 16:22:57 +0800
Subject: [PATCH 3/5] libavfilter/vf_dnn_detect: Add vaapi into detect

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>
---
 libavfilter/vf_dnn_detect.c | 32 +++++++++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

diff --git a/libavfilter/vf_dnn_detect.c b/libavfilter/vf_dnn_detect.c
index 8f50573527..319bd08663 100644
--- a/libavfilter/vf_dnn_detect.c
+++ b/libavfilter/vf_dnn_detect.c
@@ -715,6 +715,9 @@ static const enum AVPixelFormat pix_fmts[] = {
     AV_PIX_FMT_YUV420P, AV_PIX_FMT_YUV422P,
     AV_PIX_FMT_YUV444P, AV_PIX_FMT_YUV410P, AV_PIX_FMT_YUV411P,
     AV_PIX_FMT_NV12,
+#if CONFIG_VAAPI
+    AV_PIX_FMT_VAAPI,
+#endif
     AV_PIX_FMT_NONE
 };
 
@@ -838,6 +841,24 @@ static int config_input(AVFilterLink *inlink)
     return 0;
 }
 
+static int config_output(AVFilterLink *outlink)
+{
+    AVFilterContext *context = outlink->src;
+    AVFilterLink *inlink = context->inputs[0];
+    DnnDetectContext *detect_ctx = context->priv;
+    if (inlink->hw_frames_ctx) {
+        if (inlink->format == AV_PIX_FMT_VAAPI &&
+            detect_ctx->dnnctx.backend_type == DNN_OV)
+            outlink->hw_frames_ctx = av_buffer_ref(inlink->hw_frames_ctx);
+        else {
+            av_log(detect_ctx, AV_LOG_ERROR, "The dnn_backend doesn't support this pixel format\n");
+            return AVERROR_PATCHWELCOME;
+        }
+    }
+
+    return 0;
+}
+
 static const AVFilterPad dnn_detect_inputs[] = {
     {
         .name         = "default",
@@ -846,6 +867,14 @@ static const AVFilterPad dnn_detect_inputs[] = {
     },
 };
 
+static const AVFilterPad dnn_detect_outputs[] = {
+    {
+        .name         = "default",
+        .type         = AVMEDIA_TYPE_VIDEO,
+        .config_props = config_output,
+    },
+};
+
 const AVFilter ff_vf_dnn_detect = {
     .name          = "dnn_detect",
     .description   = NULL_IF_CONFIG_SMALL("Apply DNN detect filter to the input."),
@@ -853,8 +882,9 @@ const AVFilter ff_vf_dnn_detect = {
     .init          = dnn_detect_init,
     .uninit        = dnn_detect_uninit,
     FILTER_INPUTS(dnn_detect_inputs),
-    FILTER_OUTPUTS(ff_video_default_filterpad),
+    FILTER_OUTPUTS(dnn_detect_outputs),
     FILTER_PIXFMTS_ARRAY(pix_fmts),
     .priv_class    = &dnn_detect_class,
     .activate      = dnn_detect_activate,
+    .flags_internal = FF_FILTER_FLAG_HWFRAME_AWARE,
 };
-- 
2.34.1

