From 40584b3f39197987298600e45699f73417f88e77 Mon Sep 17 00:00:00 2001
From: Wenbin Chen <wenbin.chen@intel.com>
Date: Tue, 5 Mar 2024 16:22:57 +0800
Subject: [PATCH] libavfilter/vf_dnn_detect: Add vaapi into detect

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>
---
 libavfilter/vf_dnn_detect.c | 34 +++++++++++++++++++++++++++++++++-
 1 file changed, 33 insertions(+), 1 deletion(-)

diff --git a/libavfilter/vf_dnn_detect.c b/libavfilter/vf_dnn_detect.c
index 590e4e256d..078d262a78 100644
--- a/libavfilter/vf_dnn_detect.c
+++ b/libavfilter/vf_dnn_detect.c
@@ -719,6 +719,9 @@ static const enum AVPixelFormat pix_fmts[] = {
     AV_PIX_FMT_YUV420P, AV_PIX_FMT_YUV422P,
     AV_PIX_FMT_YUV444P, AV_PIX_FMT_YUV410P, AV_PIX_FMT_YUV411P,
     AV_PIX_FMT_NV12,
+#if !HAVE_WINDOWS_H && CONFIG_VAAPI
+    AV_PIX_FMT_VAAPI,
+#endif
     AV_PIX_FMT_NONE
 };
 
@@ -844,6 +847,26 @@ static int config_input(AVFilterLink *inlink)
     return 0;
 }
 
+static int config_output(AVFilterLink *outlink)
+{
+    AVFilterContext *context = outlink->src;
+    AVFilterLink *inlink = context->inputs[0];
+    DnnDetectContext *detect_ctx = context->priv;
+    FilterLink *inl = ff_filter_link(inlink);
+    FilterLink *onl = ff_filter_link(outlink);
+    if (inl->hw_frames_ctx) {
+        if (inlink->format == AV_PIX_FMT_VAAPI &&
+            detect_ctx->dnnctx.backend_type == DNN_OV)
+            onl->hw_frames_ctx = av_buffer_ref(inl->hw_frames_ctx);
+        else {
+            av_log(detect_ctx, AV_LOG_ERROR, "The dnn_backend doesn't support this pixel format\n");
+            return AVERROR_PATCHWELCOME;
+        }
+    }
+
+    return 0;
+}
+
 static const AVFilterPad dnn_detect_inputs[] = {
     {
         .name         = "default",
@@ -852,16 +875,25 @@ static const AVFilterPad dnn_detect_inputs[] = {
     },
 };
 
+static const AVFilterPad dnn_detect_outputs[] = {
+    {
+        .name         = "default",
+        .type         = AVMEDIA_TYPE_VIDEO,
+        .config_props = config_output,
+    },
+};
+
 const FFFilter ff_vf_dnn_detect = {
     .p.name        = "dnn_detect",
     .p.description = NULL_IF_CONFIG_SMALL("Apply DNN detect filter to the input."),
     .p.priv_class  = &dnn_detect_class,
+    .p.flags_internal = FF_FILTER_FLAG_HWFRAME_AWARE,
     .priv_size     = sizeof(DnnDetectContext),
     .preinit       = ff_dnn_filter_init_child_class,
     .init          = dnn_detect_init,
     .uninit        = dnn_detect_uninit,
     FILTER_INPUTS(dnn_detect_inputs),
-    FILTER_OUTPUTS(ff_video_default_filterpad),
+    FILTER_OUTPUTS(dnn_detect_outputs),
     FILTER_PIXFMTS_ARRAY(pix_fmts),
     .activate      = dnn_detect_activate,
 };
-- 
2.43.0

