From 05bc46d5931cdff56a39d3875a4f12c9e6b93c0c Mon Sep 17 00:00:00 2001
From: Wenbin Chen <wenbin.chen@intel.com>
Date: Tue, 5 Mar 2024 16:46:12 +0800
Subject: [PATCH] libavfilter/vf_dnn_classify: Add vaapi into classify

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>
---
 libavfilter/vf_dnn_classify.c | 33 ++++++++++++++++++++++++++++++++-
 1 file changed, 32 insertions(+), 1 deletion(-)

diff --git a/libavfilter/vf_dnn_classify.c b/libavfilter/vf_dnn_classify.c
index f92c41ab76..388c1d346b 100644
--- a/libavfilter/vf_dnn_classify.c
+++ b/libavfilter/vf_dnn_classify.c
@@ -199,6 +199,9 @@ static const enum AVPixelFormat pix_fmts[] = {
     AV_PIX_FMT_YUV420P, AV_PIX_FMT_YUV422P,
     AV_PIX_FMT_YUV444P, AV_PIX_FMT_YUV410P, AV_PIX_FMT_YUV411P,
     AV_PIX_FMT_NV12,
+#if !HAVE_WINDOWS_H && CONFIG_VAAPI
+    AV_PIX_FMT_VAAPI,
+#endif
     AV_PIX_FMT_NONE
 };
 
@@ -293,16 +296,44 @@ static av_cold void dnn_classify_uninit(AVFilterContext *context)
     free_classify_labels(ctx);
 }
 
+static int config_output(AVFilterLink *outlink)
+{
+    AVFilterContext *context = outlink->src;
+    AVFilterLink *inlink = context->inputs[0];
+    DnnClassifyContext *classify_ctx = context->priv;
+    FilterLink *inl = ff_filter_link(inlink);
+    FilterLink *onl = ff_filter_link(outlink);
+    if (inl->hw_frames_ctx) {
+        if (inlink->format == AV_PIX_FMT_VAAPI &&
+            classify_ctx->dnnctx.backend_type == DNN_OV)
+            onl->hw_frames_ctx = av_buffer_ref(inl->hw_frames_ctx);
+        else {
+            av_log(classify_ctx, AV_LOG_ERROR, "The dnn_backend doesn't support this pixel format\n");
+            return AVERROR_PATCHWELCOME;
+        }
+    }
+    return 0;
+}
+
+static const AVFilterPad dnn_classify_outputs[] = {
+    {
+        .name         = "default",
+        .type         = AVMEDIA_TYPE_VIDEO,
+        .config_props = config_output,
+    },
+};
+
 const FFFilter ff_vf_dnn_classify = {
     .p.name        = "dnn_classify",
     .p.description = NULL_IF_CONFIG_SMALL("Apply DNN classify filter to the input."),
     .p.priv_class  = &dnn_classify_class,
+    .flags_internal = FF_FILTER_FLAG_HWFRAME_AWARE,
     .priv_size     = sizeof(DnnClassifyContext),
     .preinit       = ff_dnn_filter_init_child_class,
     .init          = dnn_classify_init,
     .uninit        = dnn_classify_uninit,
     FILTER_INPUTS(ff_video_default_filterpad),
-    FILTER_OUTPUTS(ff_video_default_filterpad),
+    FILTER_OUTPUTS(dnn_classify_outputs),
     FILTER_PIXFMTS_ARRAY(pix_fmts),
     .activate      = dnn_classify_activate,
 };
-- 
2.43.0

